File: clients/ios/RemoteEverything/Core/Store/AtomicFile.swift
import Foundation

/// A file that is replaced, never edited: the new contents are written beside it
/// and renamed over it, so a reader sees either the whole old file or the whole
/// new one. Everything that must survive being killed halfway is written this way.
enum AtomicFile {

    static func write(_ contents: Data, to url: URL) throws {
        let directory = url.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let temporary = directory.appendingPathComponent(url.lastPathComponent + ".new")
        try contents.write(to: temporary, options: .atomic)
        if FileManager.default.fileExists(atPath: url.path) {
            // Replacing is not atomic on every filesystem, but the old file stays
            // whole until the move lands, which is what the readers depend on.
            _ = try? FileManager.default.removeItem(at: url)
        }
        try FileManager.default.moveItem(at: temporary, to: url)
    }

    static func read(_ url: URL) -> Data? {
        try? Data(contentsOf: url)
    }

    static func remove(_ url: URL) {
        try? FileManager.default.removeItem(at: url)
    }
}

File: clients/android/app/src/main/java/com/remoteeverything/core/store/AtomicFile.kt
package com.remoteeverything.core.store

import java.io.File

/**
 * A file that is replaced, never edited: the new contents are written beside it
 * and renamed over it, so a reader sees either the whole old file or the whole
 * new one. Anything that must survive being killed halfway is written this way.
 */
object AtomicFile {

    fun write(file: File, contents: ByteArray) {
        file.parentFile?.mkdirs()
        val temporary = File(file.parentFile, file.name + ".new")
        temporary.writeBytes(contents)
        if (!temporary.renameTo(file)) {
            // Windows refuses a rename onto an existing file; replacing it first is
            // still atomic enough: the old file is whole until the rename lands.
            file.delete()
            if (!temporary.renameTo(file)) {
                temporary.delete()
                throw java.io.IOException("could not replace ${file.name}")
            }
        }
    }

    fun read(file: File): ByteArray? = if (file.isFile) file.readBytes() else null

    fun delete(file: File) {
        if (file.exists() && !file.delete()) {
            throw java.io.IOException("could not delete ${file.name}")
        }
    }
}
